src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,308 @@
+import za.ac.sun.cs.green.Green;
+import za.ac.sun.cs.green.expr.Expression;
+import za.ac.sun.cs.green.service.BasicService;
+import za.ac.sun.cs.green.expr.Constant;
+import za.ac.sun.cs.green.expr.IntConstant;
+import za.ac.sun.cs.green.expr.IntVariable;
+import za.ac.sun.cs.green.expr.Operation;
+import za.ac.sun.cs.green.expr.Variable;
+import za.ac.sun.cs.green.expr.Visitor;
+import za.ac.sun.cs.green.expr.VisitorException;
+import za.ac.sun.cs.green.util.Reporter;
+
+
+public class ConstantPropogation extends BasicService {
+   private int invocations = 0;
+
+   public ConstantPropogation(Green solver) {
+       super(solver);
+   }
+
+   @Override
+   public Set<Instance> processRequest(Instance instance) {
+       @SuppressWarnings("unchecked")
+       Set<Instance> result = (Set<Instance>) instance.getData(getClass());
+       if (result == null) {
+           final Map<Variable, Variable> map = new HashMap<Variable, Variable>();
+           final Expression e = propogateConstants(instance.getFullExpression(), map);
+           final Instance i = new Instance(getSolver(), instance.getSource(), null, e);
+           result = Collections.singleton(i);
+           instance.setData(getClass(), result);
+       }
+       return result;
+   }
+
+   @Override
+	public void report(Reporter reporter) {
+		reporter.report(getClass().getSimpleName(), "invocations = " + invocations);
+   }
+public Expression propogateConstants(Expression expression, Map<Variable, Variable> map) {
+               //Use the visitor to propogate constants
+               try {
+                    log.log(Level.FINEST, "Before ConstantPropogation: " + expression);
+       			    invocations++;
+
+                    HashMap <Variable, Constant> var_map;
+
+                    MapVisitor mapVisitor = new MapVisitor();
+        			expression.accept(mapVisitor);
+        			var_map = mapVisitor.getMap();
+                    //log.log(Level.FINEST, "\n__getMap__\n " +
+                    //mapVisitor.getLogStr());
+
+                    ConstantPropogationVisitor propogationVisitor =
+                    new ConstantPropogationVisitor(var_map);
+                    expression.accept(propogationVisitor);
+                    log.log(Level.FINEST, "\n__getExpression__\n " +
+                    propogationVisitor.getLogStr());
+                    expression = propogationVisitor.getExpression();
+                    log.log(Level.FINEST, "\n__getExpression__\n " +
+                    propogationVisitor.getLogStr());
+
+                    log.log(Level.FINEST, "After ConstantPropogation: " + expression);
+        			return expression;
+               } catch (VisitorException x) {
+                   log.log(Level.SEVERE,
+       					"encountered an exception -- this should not be happening!",
+       					x);
+               }
+               return null;
+           }
+
+/*
+ * Goal is to obtain a HashMap of variables and constant values assigned in the
```

Comment:
I'm glad to see some comments, but Javadoc is the preferred way of commenting on a top level thing like a class or a method.